pkg/disttask/framework/scheduler/balancer_test.go

```diff
@@ -224,6 +225,44 @@ func TestBalanceOneTask(t *testing.T) {
 			},
 			expectedUsedSlots: map[string]int{"tidb2": 16, "tidb3": 16},
 		},
+		// balanced, but max node count is limited.
+		{
+			subtasks: []*proto.SubtaskBase{
+				{ID: 1, ExecID: "tidb1", Concurrency: 16, State: proto.SubtaskStateRunning},
+				{ID: 2, ExecID: "tidb2", Concurrency: 16, State: proto.SubtaskStateRunning},
+				{ID: 3, ExecID: "tidb2", Concurrency: 16, State: proto.SubtaskStatePending},
+			},
+			maxNodeCount:  1,
+			eligibleNodes: []string{"tidb1", "tidb2"},
+			initUsedSlots: map[string]int{"tidb1": 0, "tidb2": 0},
+			expectedSubtasks: []*proto.SubtaskBase{
+				{ID: 1, ExecID: "tidb2", Concurrency: 16, State: proto.SubtaskStateRunning},
+				{ID: 2, ExecID: "tidb2", Concurrency: 16, State: proto.SubtaskStateRunning},
+				{ID: 3, ExecID: "tidb2", Concurrency: 16, State: proto.SubtaskStatePending},
+			},
+			expectedUsedSlots: map[string]int{"tidb1": 0, "tidb2": 16},
+		},
+		// scale out, but max node count is limited.
+		{
+			subtasks: []*proto.SubtaskBase{
+				{ID: 1, ExecID: "tidb1", Concurrency: 16, State: proto.SubtaskStateRunning},
+				{ID: 2, ExecID: "tidb1", Concurrency: 16, State: proto.SubtaskStatePending},
+				{ID: 3, ExecID: "tidb1", Concurrency: 16, State: proto.SubtaskStatePending},
+				{ID: 4, ExecID: "tidb1", Concurrency: 16, State: proto.SubtaskStatePending},
+				{ID: 5, ExecID: "tidb1", Concurrency: 16, State: proto.SubtaskStatePending},
```

Comment:
maybe make them on `tidb3` initially. they we are sure `tidb3` is ordered first

now `tidb1` is already the first in `eligibleNodes`

Reply:
https://github.com/pingcap/tidb/pull/58937/commits/1fd36193024b09a3c859ea972a525cdedac5b3bc